### ISSUE TITLE
EPUB main spec

### ISSUE THREAD

**Issue by sujato** [suttacentral/suttacentral#2364]
We should output valid and well-formed EPUBs of publications. These will typically contain similar content to the other versions, but optimized for the EPUB format. 

As of writing, the latest EPUB spec is 3.2, so let us target that.  

EPUBs are essentially HTML with a container and some metadata. Thus conversion to EPUB from HTML is trivial. The main issue is getting the details of content right.

For the conversion, probably pandoc will work?

Check out the "Standard Ebooks" project: they do a great job of focusing on quality output. Let's make ours just as good!

https://standardebooks.org/

Here's their detailed and super-useful spec. Let's follow it where applicable:

https://standardebooks.org/manual/1.6.3/single-page

Also see:

https://readium.org/awesome-readium/

## process

- define EPUB in `editions`
- create standalone HTML from config
- convert HTML to EPUB
- make sure cover image is included.

## file size

Generally we will make an EPUB of a single nikaya or equivalent. This is comparable in size to a large novel, say *War & Peace*, at up to 500,000 words. 

There is a recommendation for EPUBs to contain files less than 300kb, to ensure that all devices will support them. So we should split the HTML file.

As a sensible default, let's split the file at the *vagga* ("Chapter") level. In HTML, this will be the heading level immediately above the `sutta-title`, whatever that is.

In some cases—notably DN and Khandhakas—splitting at the individual sutta level would be preferable. Perhaps this could be an option.

## fonts

EPUBs technically support embedded fonts. However support is flaky, they add a lot to file size, users frequently override them, and most of the main devices have reasonable fonts with good enough character support. So I think embedding fonts will be more hassle than it's worth.

## ToC

There are two ways to make a ToC in EPUBs:

- [the EPUB native way](https://www.w3.org/publishing/epub3/epub-packages.html#sec-package-nav)
- linked list in HTML

We use the EPUB standard for the main-toc, and HTML lists for secondary-toc. 

### flatten headings

Note that certain ereaders only support main-toc to a depth of two. This includes the Kindle; even though we don't directly target Kindle (because it uses a proprietary format) we should support conversion to it. And there may well be other similar cases.

Now, normally a main-toc will have 1-3 levels. In fact, of the major collections, only MN will have three levels. Let's flatten these for EPUB so we get around this issue.

In the flattened file, the lowest `heading` level (which will be `sutta-title`) becomes h2, and all levels above that become h1. Levels below that (i.e. the `subheadings`) will be upped to start with h3.

## comments

See #2364 

## styles

EPUBs theoretically support CSS, but writing CSS for EPUBs is like saving a web page that was made in React, opening it in Word, saving as .docx, using a random web service to convert it back to HTML, opening in IE5, and wondering, "you mean it doesn't support CSS `grid`?"

Here are some general guidelines. (Thanks @thesunshade)

- **Don't use colour to indicate anything.** Expect users to be using e-ink where it's only grayscale. Especially don't use any "off black" for any reason as it just decreases the contrast unless the goal really is to make something less readable, e.g. verse numbers.
- **Work in em measurements.** This is good practice generally. The main body text should be 1em
- **Do style headings to an appropriate size.** Devices/software generally have terrible default heading sizes. These should definitely not be left up to the device to decide. It's possible that h1 may be by default twice the body font size, ending up with a title that fills the entire screen. 
- **Set headings in regular weight.** By default headings will be bolded. In practice this almost always looks bad, and is also unnecessary if the font size is large. The only time I will have bold headings is if a heading font-size has to get down to the same size as paragraph text.

Try this to start.

```
body{
font-size: 1em;
}
#title-page{
    text-align:  center;
    margin-bottom:  8em;
}
.translation_title{
    font-size:  3em;
    line-height:  1.2;
    margin:  2em 0 1em;
}
.translation_subtitle{
    font-size:  1.5em;
    line-height:  1.2;
    margin:  2em 0 10em;
}
.author_name{
    font-size:  1.5em;
    font-style:  italic;
    margin:  2em 0;
}
#license{
    padding:  1em 2em 2em;
    font-size:  0.8em;
    border:  1px solid #ccc;
}
h1{
font-size: 1.5em;
font-weight: normal;
text-align: center;
line-height: 1.3333;
margin: 4em 0 2em 0;
}
h2{
font-size: 1.333em;
font-weight: normal;
text-align: left;
line-height: 1.3333;
}
h3{
font-size: 1.2em;
font-weight: normal;
text-align: left;
line-height: 1.3333
}
h4{
font-size: 1em;
font-weight: bold;
text-align: left;
}
h5{
font-size: 1em;
font-weight: normal;
font-style: italic;
text-align: left;
}
h6{
font-size: 1em;
font-weight: normal;
font-style: normal;
text-align: left;
}
h2, h3,  h4,  h5,  h6 {
    margin: 1.5em 0 0 0;
  }
  .sutta-title{
    text-align:  center;
    margin: 4em 0 2em 0;
  }
  p, p + ul, p + ol, p + dl {
    margin: 1em 0 0 0;
  }
```

**Comment by sparkbold** [comment]
Look like the server is down https://suttacentral.net/api/ebook/dn_en_sujato.epub
{message: "InternalServer Error"}

**Comment by sujato** [comment]
Yeah that is the old implementation, we should remove it. This replaces it.

**Comment by thesunshade** [comment]
> EPUBs theoretically support CSS, but writing CSS for EPUBs is like

Umm... I'm not sure I follow? In any case, for the publications you are working on, I think the css is solid. It has more to do with the device/rendering software. Speaking of which, it use to be that the rendering software on most/all epub devices was actually made by Adobe, so testing them on the desktop version, Adobe Digital Editions, was considered best practice.

Tables can be tricky, but I think that's mostly because of the small form factor added to the fact that the device is really thinking about "pages" in how the user interacts, and tables going across pages can be messy. 

```css
body{
font-size: 1em;
line-height:  1.5;
}
```
I recommend against messing with the line height. This is something that the user can often control through the interface. I know if I opened a book and suddenly the line height was larger (or smaller) than all my other books I would be annoyed. It's true that increasing line height improves readability. However when reading on a small device it also means that you are getting less text on the page and thus more frequent page turns. 

Really anything that has to do with readability is handled by the device. After all, they are designed for reading. 

The best resource I've found on the web is https://www.mobileread.com/forums/forumdisplay.php?s=dadd7abeed4d8c4d6bbf2b2300559391&f=179. Lots of experts on the epub format there. There are plenty of other forums, but that's the one that deals directly with the format.

**Comment by sujato** [comment]
I don't have any attachment to setting line-height, but in the first app I opened it in (calibre) the line height was cramped, so I thought a sensible default might be in order.

**Comment by thesunshade** [comment]
I posted the question of setting line height to that forum and the consensus seemed to be to not set line height. Other than the fact that it would annoy people who had already set the line height through the device, one person explained that since people could change the *font* they used, it was best left to the device to decide the proper line height based on the font metrics. 

> The sense I get is that most of the epubs around the place are so poorly formed, even a plain but semantically correct file will work just fine.

Yes, this is probably a good conclusion. Metta Forest Monastery seems to put some effort into making good looking epubs. There is also [Standard Ebooks](https://standardebooks.org/contribute) that prides themselves on well designed ebooks. But really, once you start reading, even a poorly designed book is usually fine.

> Do you know how large file sizes affect performance?

My understanding is that devices treat ebooks as the folder of files that they are. The epub standard has a limit to the size of the individual xhtml files, but in practice most people make each chapter its own file so you never come close to that limit. All of that is to say that no, file size does not affect performance. Each individual xhtml file is loaded into memory as needed.

> Should we offer nikayas in volumes of EPUBs? Or alternatively, I wonder if there's any performance difference in making the epub from one big file (as currently) or many small files (say one per sutta)?

It's kind of a Goldilocks problem. You don't want them too big and you don't want them too small.

It's helpful to have an understanding of how people are going to use the ebooks. One of the reasons that early ebooks were so bad is that they were designed by people who either never read an ebook or they actively hated the idea. 

Ebooks are really meant for reading. Yes, you can do searches. Yes, you can use links. But the thing they are really good for is reading. The other thing they are really good at is being small and portable. But for doing research, skipping around place to place, they really aren't even as good a paper book. 

Another thing to keep in mind is that not all ebook devices or software do a great job of managing collections of books. This is because most users are not keeping a precious library on their device. They are checking out books from the library, reading the latest best seller, or the next book in their favorite genre series. 

Unfortunately the Kindle is probably the worst at this because it doesn't allow grouping folders inside of folders. Therefore having each sutta be its own book would be an organizational disaster. Even if the device/software allowed it, it would still be cumbersome. 

But then the other limitation you face is how the table of contents should be structured. That's why it's important to make a table of contents effectively since books like SN and AN will contain lots of suttas. 

My opinion is that if you are going to have standard books you generate as epubs, they should be in the traditional groupings you have now, one book for each nikaya, except KN where the individual books will each be their own epub.

Now, Metta Forest does produce a single ebook of all of Ajahn T's sutta translations, but the only reason for that is that he does lots of linking between suttas across the canon. It's 5mb, which is huge for an ebook. But it loads just fine. All that is just as an FYI.

Looking at that book I'm reminded of something else important in my opinion for the table of contents. Each TOC entry should start with the sutta number. Because that is how some (many?) people will be looking up suttas. Putting the samyutta number or the AN chapter number may not be necessary.

**Comment by sujato** [comment]
Thanks!

I realize my question was not clearly phrased!

What I meant was, currently our nikaya EPUBs are formed from a single HTML file, and so the EPUB file is a single file with chapters delineated by headings only, not by file. Let's say a nikaya is up to 500,000 words, comparable to War and Peace.

That's the easiest way to produce them: we form the structure using the heading levels in HTML and just pandoc it to make an EPUB.

Here's DN for example. (Zipped to make it allowable for Github.)

[dn-bare.zip](https://github.com/suttacentral/suttacentral/files/7953792/dn-bare.zip)

So what I was wondering is whether we should be splitting the HTML file within the EPUB to be, say, individual vaggas or suttas or whatever.

But anyway, you mentioned about a file size limit. You say it's in the standard: is that correct? I can't find it there? Threads discussing this speak of it as a device limitation or a recommendation. In any case, probably best to go ahead and split the files by vagga or something.

**Comment by thesunshade** [comment]
> they don't bundle fonts.

Personally, I have never, ever liked a publisher font more than whatever the latest default Kindle font was. I think e-ink is just so different from both paper and regular screens that you're unlikely to improve on what the device designers include.

And apologies for my too long answer to a question you didn't ask! As well, sorry for quoting what appears to be a non-existent standard. I'll be more careful in the future. It may have been an old mobi thing since that was originally for palm pilots years ago.

I've always done my books with one sutta per file, even on SN and AN. An added benefit to that is that it guarantees a page break before each sutta. You should be able to do that through css, though. It's a style decision, but I find that with small screens it's always nice to start suttas each on a new page, something you'd never do with print books. In theory you could use widow/orphan control, but I'm not sure how well that is supported.

**Comment by thesunshade** [comment]
I don't know if this will have an impact on the books you will be producing, but I just discovered that `p span:lang(pli)` is not recognized by many epub reader apps. I have tested 4 so far that do not work. ADE, Google Play Books seems to work. But Freda, Prestigo, Media365 and one other do not. 

It passes validation, but whatever renderer these free apps are using don't seem to support it. 
¯\\\_(ツ)_/¯
